### ISSUE TITLE
json-stream tokenizer unicode surrogate handling

### ISSUE THREAD

**Issue by goinnn** [daggaz/json-stream#39]
json-stream should accept byte string because we can not parse some JSON with json-stream but we can parse them with json library.

Using json library works:
 
    import json
    json_bytes = b'{"test": "Direkt auf dem Parkplatz der Gastwirtschaft Neugebauer zu finden, 24/7 \xc3\xb6ffentlich Zug\xc3\xa4nglich,\\r\\nLeckere Spei\xc3\x9fen und Getr\xc3\xa4nke von Donnerstag bis S \\uD83D\\uDE99"}'
    import json
    json.loads(json_bytes)['test']
    
    Result: 'Direkt auf dem Parkplatz der Gastwirtschaft Neugebauer zu finden, 24/7 öffentlich Zugänglich,\r\nLeckere Speißen und Getränke von Donnerstag bis S 🚙'

Using json_stream library does not  work:

     import json_stream
     from io import StringIO
     json_bytes = b'{"test": "Direkt auf dem Parkplatz der Gastwirtschaft Neugebauer zu finden, 24/7 \xc3\xb6ffentlich Zug\xc3\xa4nglich,\\r\\nLeckere Spei\xc3\x9fen und Getr\xc3\xa4nke von Donnerstag bis S \\uD83D\\uDE99"}'
     json_stream.load(StringIO(json_bytes.decode('utf-8')))['test']

     raise a exception (ValueError: No unicode character for code: 55357 at index 162)

We can do something like this:

     import json_stream
     from io import StringIO
     json_bytes = b'{"test": "Direkt auf dem Parkplatz der Gastwirtschaft Neugebauer zu finden, 24/7 \xc3\xb6ffentlich Zug\xc3\xa4nglich,\\r\\nLeckere Spei\xc3\x9fen und Getr\xc3\xa4nke von Donnerstag bis S \\uD83D\\uDE99"}'
     json_stream.load(StringIO(json_bytes.decode('utf-8').encode('latin-1').decode('unicode-escape').encode('utf-16', 'surrogatepass').decode('utf-16')))['test']

     Result: 'Direkt auf dem Parkplatz der Gastwirtschaft Neugebauer zu finden, 24/7 öffentlich Zugänglich,\r\nLeckere Speißen und Getränke von Donnerstag bis S 🚙'

But this does not work with another byte strings:

     import json_stream
     from io import StringIO
     json_bytes = b'{"test": "\xc4\x8d \\uD83D\\uDE99"}'
     json_stream.load(StringIO(json_bytes.decode('utf-8').encode('latin-1').decode('unicode-escape').encode('utf-16', 'surrogatepass').decode('utf-16')))['test']

    raise a exception UnicodeEncodeError: 'latin-1' codec can't encode character '\u010d' in position 10: ordinal not in range(256)

But this works perfectly with json library:

    import json
    json_bytes = b'{"test": "\xc4\x8d \\uD83D\\uDE99"}'
    json.loads(json_bytes)['test']
   
    Result: 'č 🚙'

Also there is another relational problem. The original json_bytes variable contained quotation marks:

    json_bytes = b'{"test": "Direkt auf dem Parkplatz der Gastwirtschaft \\"Neugebauer\\" zu finden, 24/7 \xc3\xb6ffentlich Zug\xc3\xa4nglich,\\r\\nLeckere Spei\xc3\x9fen und Getr\xc3\xa4nke von Donnersta bis S \\uD83D\\uDE99"}'

If I run the solution, json-stream cut the result:

    import json_stream
    from io import StringIO
    json_bytes = b'{"test": "Direkt auf dem Parkplatz der Gastwirtschaft \\"Neugebauer\\" zu finden, 24/7 \xc3\xb6ffentlich Zug\xc3\xa4nglich,\\r\\nLeckere Spei\xc3\x9fen und Getr\xc3\xa4nke von Donnersta bis S \\uD83D\\uDE99"}'
    json_stream.load(StringIO(json_bytes.decode('utf-8').encode('latin-1').decode('unicode-escape').encode('utf-16', 'surrogatepass').decode('utf-16')))['test']

    Result: 'Direkt auf dem Parkplatz der Gastwirtschaft '

**Comment by smheidrich** [comment]
Thanks for the issue! Let me try to disentangle that a bit:

The byte string thing is a red herring. All these issues happen with regular Python strings as well, because AFAICT the actual bug is improper handling of chains of multiple `\u...` escape sequences in JSON.

This snippet produces the same exception saying `ValueError: No unicode character for code: 55356 at index 15` without any byte strings involved:

```python
from io import StringIO
from json_stream import load

json_str = r'{"test": "\uD83C\uDFD4\uFE0F"}'
load(StringIO(json_str))["test"]
```

Single Unicode escape sequences work fine:

```python
json_str = r'{"test": "\u00C4"}'
print(json_stream.load(StringIO(json_str))["test"])  # outputs 'Ä'
```

Why is that? The underlying issue is that the JSON standard mandates that [*if* one decides to escape Unicode code points beyond `U+FFFF`, it should be done using UTF-16 surrogate pair sequences](https://stackoverflow.com/q/38463038), but json-stream's Rust tokenizer naively tries to [parse each `\u...` escape into a Rust char separately](https://github.com/smheidrich/py-json-stream-rs-tokenizer/blob/89156c9f512b73f1a9776743e06feae1b2158bd5/src/lib.rs#L659), and [Rust chars specifically can't represent Unicode surrogate code points](https://doc.rust-lang.org/std/primitive.char.html#validity). So that has to be fixed.

json-stream's pure-Python tokenizer appears to be able to parse them without raising an exception, but Python doesn't really seem to like the strings this produces:

```python
from io import StringIO
from json_stream.tokenizer import load, tokenize

json_str = '{"test": "\\uD83C\\uDFD4\\uFE0F"}'
result = load(StringIO(json_str), tokenizer=tokenize)['test']
print(result)
```

This will raise `UnicodeEncodeError: 'utf-8' codec can't encode characters in position 0-1: surrogates not allowed` in the `print(result)` line. I'm not sure whether these are technically correct or not, because the JSON standard does say

>However, whether a processor of JSON texts interprets such a surrogate pair as a single code point or as an
explicit surrogate pair is a semantic decision that is determined by the specific processor.

But I guess if Python strings, being pure UTF-8, can't contain these, then we should convert them to their UTF-8 equivalent anyway. So that should probably be fixed as well.

Finally, @goinnn, for your last point about quotation marks, that seems to be an issue with that wild chain of decoding and encoding you're doing rather than with json-stream:

```python
>>> json_bytes = b'{"test": "Direkt auf dem Parkplatz der Gastwirtschaft \\"Neugebauer\\" zu finden, 24/7 \xc3\xb6ffentlich Zug\xc3\xa4nglich,\\r\\nLeckere Spei\xc3\x9fen und Getr\xc3\xa4nke von Donnersta bis S \\uD83D\\uDE99"}'
>>> json_bytes.decode('utf-8').encode('latin-1').decode('unicode-escape').encode('utf-16', 'surrogatepass').decode('utf-16'))
'{"test": "Direkt auf dem Parkplatz der Gastwirtschaft "Neugebauer" zu finden, 24/7 öffentlich Zugänglich,\r\nLeckere Speißen und Getränke von Donnersta bis S 🚙"}'
```

The double quotes inside the JSON string lost the backslash that escaped them, so the JSON string does in fact end before `Neugebauer` then. Since json-stream yields that string before parsing further, it will only raise an error due to the now-malformed JSON if you request the whole dictionary, e.g. by doing `list(load(...).items())` instead of `load(...)["test"]`.

**Comment by goinnn** [comment]
Thanks @smheidrich, so are there three bugs?

**Comment by smheidrich** [comment]
@goinnn Just one that's relevant for you, which should be fixed in `json-stream-rs-tokenizer >= 0.4.17`, so you could try updating and seeing if it works now.

**Comment by goinnn** [comment]
Hi @smheidrich,

**Now works in our cases**. Thanks!! 🥳🥳

```python
import json
import json_stream
from io import StringIO
json_bytes = b'{"test": "Direkt auf dem Parkplatz der Gastwirtschaft Neugebauer zu finden, 24/7 \xc3\xb6ffentlich Zug\xc3\xa4nglich,\\r\\nLeckere Spei\xc3\x9fen und Getr\xc3\xa4nke von Donnerstag bis S \\uD83D\\uDE99"}'
json_stream.load(StringIO(json_bytes.decode('utf-8')))['test']

Result: 'Direkt auf dem Parkplatz der Gastwirtschaft Neugebauer zu finden, 24/7 öffentlich Zugänglich,\r\nLeckere Speißen und Getränke von Donnerstag bis S 🚙'

json_stream.load(StringIO(json_bytes.decode('utf-8')))['test'] == json.loads(json_bytes)['test']
Result: True
```


```python
import json
import json_stream
from io import StringIO
json_bytes = b'{"test": "\xc4\x8d \\uD83D\\uDE99"}'
json_stream.load(StringIO(json_bytes.decode('utf-8')))['test']

Result: 'č 🚙'

json_stream.load(StringIO(json_bytes.decode('utf-8')))['test'] == json.loads(json_bytes)['test']
Result: True
```

```python
import json
import json_stream
from io import StringIO
json_bytes = b'{"test": "Direkt auf dem Parkplatz der Gastwirtschaft \\"Neugebauer\\" zu finden, 24/7 \xc3\xb6ffentlich Zug\xc3\xa4nglich,\\r\\nLeckere Spei\xc3\x9fen und Getr\xc3\xa4nke von Donnersta bis S \\uD83D\\uDE99"}'
json_stream.load(StringIO(json_bytes.decode('utf-8')))['test']

Result: 'Direkt auf dem Parkplatz der Gastwirtschaft "Neugebauer" zu finden, 24/7 öffentlich Zugänglich,\r\nLeckere Speißen und Getränke von Donnersta bis S 🚙'

json_stream.load(StringIO(json_bytes.decode('utf-8')))['test'] == json.loads(json_bytes)['test']
Result: True
```

**Comment by daggaz** [comment]
@smheidrich perhaps there are things you could take from #42 that will improve the rest tokenizer?

**Comment by smheidrich** [comment]
@daggaz I'll probably do just that, as I also thought that needing 4 different states, one for each digit of a Unicode codepoint, seems excessive, but decided against refactoring it to maintain some similarity to the Python tokenizer. With that barrier gone, I'll probably copy your approach :) Thanks.

**Comment by daggaz** [comment]
this change to the python tokenizer has now been merged, and a new release 2.3.1 will be available shortly
